htmresearch/algorithms/sparse_net.py

```diff
@@ -0,0 +1,352 @@
+    if len(data.shape) == 1:
+      data = data[:, np.newaxis]
+
+
+    projection = self.basis.T.dot(data)
+    representation = self.basis.T.dot(self.basis) - np.eye(self.outputDim)
+    states = np.zeros((self.outputDim, data.shape[1]))
+
+    threshold = 0.5 * np.max(np.abs(projection), axis=0)
+    activations = self._activation(states, threshold)
+
+    for _ in xrange(self.numLcaIterations):
+      # update dynamic system
+      states *= (1 - self.lcaLearningRate)
+      states += self.lcaLearningRate * (projection - representation.dot(activations))
+      activations = self._activation(states, threshold)
+
+      # decay threshold
+      threshold *= self.thresholdDecay
+      threshold[threshold < self.minThreshold] = self.minThreshold
+
+    return activations
+
+
+  def plotLoss(self, filename=None):
+    """
+    Plots the loss history.
+
+    :param filename    (string) Can be provided to save the figure
+    """
+    plt.figure()
+    plt.plot(self.losses.keys(), self.losses.values())
+    plt.xlabel("Iteration")
+    plt.ylabel("Loss")
+    plt.title("Learning curve for {}".format(self))
+
+    if filename is not None:
+      plt.savefig(filename)
+
+
+  def plotBasis(self, filename=None):
+    """
+    Plots the basis functions, reshaped in 2-dimensional arrays.
+
+    This representation makes the most sense for visual input.
+    :param:  filename    (string) Can be provided to save the figure
+    """
+    if np.floor(np.sqrt(self.inputDim)) ** 2 != self.inputDim:
+      print "Basis visualization is not available if inputDim is not a square."
+      return
+
+    dim = int(np.sqrt(self.inputDim))
+
+    if np.floor(np.sqrt(self.outputDim)) ** 2 != self.outputDim:
+      outDimJ = np.sqrt(np.floor(self.outputDim / 2))
+      outDimI = np.floor(self.outputDim / outDimJ)
+      if outDimI > outDimJ:
+        outDimI, outDimJ = outDimJ, outDimI
+    else:
+      outDimI = np.floor(np.sqrt(self.outputDim))
+      outDimJ = outDimI
+
+    outDimI, outDimJ = int(outDimI), int(outDimJ)
+    basis = - np.ones((1 + outDimI * (dim + 1), 1 + outDimJ * (dim + 1)))
+
+    # populate array with basis values
+    k = 0
+    for i in xrange(outDimI):
+      for j in xrange(outDimJ):
+        colorLimit = np.max(np.abs(self.basis[:, k]))
+        mat = np.reshape(self.basis[:, k], (dim, dim)) / colorLimit
+        basis[1 + i * (dim + 1) : 1 + i * (dim + 1) + dim, \
+              1 + j * (dim + 1) : 1 + j * (dim + 1) + dim] = mat
+        k += 1
+
+    plt.figure()
+    plt.subplot(aspect="equal")
+    plt.pcolormesh(basis)
+
+    plt.axis([0, 1 + outDimJ * (dim + 1), 0, 1 + outDimI * (dim + 1)])
+    # remove ticks
+    plt.gca().xaxis.set_major_locator(plt.NullLocator())
+    plt.gca().yaxis.set_major_locator(plt.NullLocator())
+    plt.title("Basis functions for {0}".format(self))
+
+    if filename is not None:
+      plt.savefig(filename)
+
+
+  def _reset(self):
```

Comment:
In addition to reset, it might be worth adding read/write serialization functionalities. We may need to save the basis and use it later.